palette/moreland/luminance.go

```diff
@@ -0,0 +1,263 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package moreland
+
+import (
+	"fmt"
+	"image/color"
+	"math"
+
+	"github.com/gonum/plot/palette"
+)
+
+// luminance is a color palette that interpolates
+// between control colors in a way that ensures a linear relationship
+// between the luminance of a color and the value it represents.
+type luminance struct {
+	colors  []cieLAB
+	scalars []float64
+
+	// alpha represents the opacity of the returned
+	// colors in the range (0,1). It is set to 1 by default.
+	alpha float64
+
+	// min and max are the minimum and maximum values of the range of scalars
+	// that can be mapped to colors using this ColorMap.
+	min, max float64
+}
+
+// NewLuminance creates a new Luminance ColorMap from the given controlColors.
+// luminance is a color palette that interpolates
+// between control colors in a way that ensures a linear relationship
+// between the luminance of a color and the value it represents.
+// If the luminance of the controls is not monotonically increasing, an
+// error will be returned.
+func NewLuminance(controls []color.Color) (palette.ColorMap, error) {
+	l := luminance{
+		colors:  make([]cieLAB, len(controls)),
+		scalars: make([]float64, len(controls)),
+		alpha:   1,
+	}
+	max := math.Inf(-1)
+	min := math.Inf(1)
+	for i, c := range controls {
+		lab := colorTosRGBA(c).cieLAB()
+		l.colors[i] = lab
+		max = math.Max(max, lab.L)
+		min = math.Min(min, lab.L)
+		if i > 0 && lab.L <= l.colors[i-1].L {
+			return nil, fmt.Errorf("moreland: luminance of color %d (%g) is not "+
+				"greater than that of color %d (%g)", i, lab.L, i-1, l.colors[i-1].L)
+		}
+	}
+	// Normalize scalar values to the range (0,1).
+	rnge := max - min
+	for i, c := range l.colors {
+		l.scalars[i] = (c.L - min) / rnge
+	}
+	// Avoid floating point math problems.
+	l.scalars[0] = 0
+	l.scalars[len(l.scalars)-1] = 1
+	return &l, nil
+}
+
+// At implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) At(scalar float64) (color.Color, error) {
```

Comment:
s/scalar/v/g and make the line below `scalar := (v - l.min) / l.max`, Also in smoothDiverging.